src/transit_data/data_init.rs

```diff
@@ -46,15 +46,133 @@ use crate::{
 
+#[derive(Clone, Debug)]
+pub(super) enum StayInType {
+    SameStopPoint(BaseVehicleJourneyIdx),
+    DifferentStopPoint(BaseVehicleJourneyIdx),
+}
+
+struct VJGroupedByStayIn {
+    pub vehicle_journey_to_prev_stay_in: HashMap<BaseVehicleJourneyIdx, StayInType>,
+    pub vehicle_journey_to_next_stay_in: HashMap<BaseVehicleJourneyIdx, StayInType>,
+}
+
+impl VJGroupedByStayIn {
+    pub fn new(base_model: &BaseModel) -> Self {
+        // the HashMap 'stay_in_vj' is used to group vehicle_journey with the same 'blocked_id' and timezone
+        // allowing use to use a vehicle of this group and move to another vehicle of the same group without transfer
+        let mut stay_in_vj = HashMap::new();
+
+        // Fill stay_in_vj
+        for vehicle_journey_idx in base_model.vehicle_journeys() {
+            let vehicle_journey = base_model.vehicle_journey(vehicle_journey_idx);
+            let block_id = &vehicle_journey.block_id;
+            let timezone = base_model.timezone(vehicle_journey_idx);
+
+            if let (Some(block_id), Some(timezone)) = (block_id, timezone) {
+                if let Ok(stop_times) = base_model.stop_times(vehicle_journey_idx) {
+                    // for now, we do not want to have stay_in on vehicle journeys
+                    // with multiple local zone
+                    // !todo find a better way to check for multiple local zone
+                    let mut local_zones: Vec<_> =
+                        stop_times.clone().map(|s| s.local_zone_id).collect();
+                    local_zones.sort_unstable();
+                    local_zones.dedup();
+
+                    if !block_id.is_empty() && stop_times.len() > 0 && local_zones.len() == 1 {
+                        let vehicle_journeys_idx = stay_in_vj
+                            .entry((block_id.as_str(), timezone))
+                            .or_insert_with(Vec::new);
+                        vehicle_journeys_idx.push(vehicle_journey_idx);
+                    }
+                }
+            }
+        }
+
+        // Sort stay_in_vj
+        for vec_idx in stay_in_vj.values_mut() {
+            vec_idx.sort_unstable_by(|lhs_idx, rhs_idx| {
+                let lhs_first_stoptime = base_model
+                    .stop_times(*lhs_idx)
+                    .unwrap() // unwrap is safe because already checked in fn new
+                    .next()
+                    .unwrap() // unwrap is safe because already check in fn new stop_times.len() > 0
+                    .board_time;
+                let rhs_first_stoptime = base_model
+                    .stop_times(*rhs_idx)
+                    .unwrap() // unwrap is safe because already checked in fn new
+                    .next()
+                    .unwrap() // unwrap is safe because already check in fn new stop_times.len() > 0
+                    .board_time;
+                lhs_first_stoptime.cmp(&rhs_first_stoptime)
+            });
```

Comment:
I think this is equivalent to
```suggestion
            vec_idx.sort_unstable_by_key(|vehicle_journey_idx| {
                base_model
                    .stop_times(*vehicle_journey_idx)
                    .unwrap() // unwrap is safe because already checked in fn new
                    .next()
                    .unwrap() // unwrap is safe because already check in fn new stop_times.len() > 0
                    .board_time;
            });
```